File: 2021/python/day06.py
from collections import defaultdict
F = defaultdict(int)
for n in open("../input/day06.input", encoding="utf-8").read().strip().split(","):
    F[int(n)] += 1

for day in range(256):
    new, FNext = F[0], {i: F[(i + 1) % 9] for i in range(9)}
    FNext[6] += new
    F = FNext
    if day + 1 == 80:
        print("Answer A:", sum(F.values()))
print("Answer B:", sum(F.values()))

File: 2021/golang/day06.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func solveDay6(fish []string, days int) int {
	sum, F := 0, make(map[int]int)
	for _, f := range fish {
		ix, _ := strconv.Atoi(f)
		F[ix]++
	}
	for i := 1; i <= days; i++ {
		spawns, Fnew := F[0], make(map[int]int)
		for j := 0; j < 8; j++ {
			Fnew[j] = F[j+1]
		}
		Fnew[6], Fnew[8] = Fnew[6]+spawns, F[0]
		F = Fnew
	}
	for _, v := range F {
		sum += v
	}
	return sum
}

func Day06() {
	b, _ := ioutil.ReadFile("../input/day06.input")
	fmt.Println("Answer A:", solveDay6(strings.Split(string(b), ","), 80))
	fmt.Println("Answer B:", solveDay6(strings.Split(string(b), ","), 256))
}
